src/target/puya.c

```diff
@@ -0,0 +1,171 @@
+// (yes, this *does* include undocumented free extra flash/ram in the 002A)
+#define FLASH_RAM_SZ 0x1fff0ffcU
+
+#define FLASH_BASE      0x40022000U
+#define FLASH_KEYR      (FLASH_BASE + 0x008U)
+#define FLASH_KEYR_KEY1 0x45670123U
+#define FLASH_KEYR_KEY2 0xcdef89abU
+
+#define FLASH_SR     (FLASH_BASE + 0x010U)
+#define FLASH_SR_BSY (1U << 16U)
+
+#define FLASH_CR        (FLASH_BASE + 0x014U)
+#define FLASH_CR_LOCK   (1U << 31U)
+#define FLASH_CR_PGSTRT (1U << 19U)
+#define FLASH_CR_PER    (1U << 1U)
+#define FLASH_CR_PG     (1U << 0U)
+
+#define FLASH_TS0     (FLASH_BASE + 0x100U)
+#define FLASH_TS1     (FLASH_BASE + 0x104U)
+#define FLASH_TS2P    (FLASH_BASE + 0x108U)
+#define FLASH_TPS3    (FLASH_BASE + 0x10cU)
+#define FLASH_TS3     (FLASH_BASE + 0x110U)
+#define FLASH_PERTPE  (FLASH_BASE + 0x114U)
+#define FLASH_SMERTPE (FLASH_BASE + 0x118U)
+#define FLASH_PRGTPE  (FLASH_BASE + 0x11cU)
+#define FLASH_PRETPE  (FLASH_BASE + 0x120U)
+
+/* RAM */
+#define RAM_START 0x20000000U
+
+/* RCC */
+#define RCC_BASE  0x40021000U
+#define RCC_ICSCR (RCC_BASE + 0x04U)
+
+/* DBG */
+#define DBG_BASE   0x40015800U
+#define DBG_IDCODE (DBG_BASE + 0x00U)
+
+/*
+ * Flash functions
+ */
+static bool puya_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool puya_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+static bool puya_flash_prepare(target_flash_s *flash);
+static bool puya_flash_done(target_flash_s *flash);
+
+bool puya_probe(target_s *target)
+{
+	uint32_t ram_size = 0U;
+	size_t flash_size = 0U;
+
+	const uint32_t dbg_idcode = target_mem32_read32(target, DBG_IDCODE);
+	if ((dbg_idcode & 0xfff) == 0) {
+		const uint32_t flash_ram_sz = target_mem32_read32(target, FLASH_RAM_SZ);
+		flash_size = ((flash_ram_sz & 7) + 1) * 0x2000;
+		ram_size = (((flash_ram_sz >> 4) & 3) + 1) * 0x800;
+		// TODO: which part families does this actually correspond to?
+		// Tested with a PY32F002AW15U which returns 0x60001000 in IDCODE
+		target->driver = "PY32Fxxx";
+	} else {
+		DEBUG_TARGET("Unknown PY32 device %08" PRIx32 "\n", dbg_idcode);
+		return false;
+	}
+
+	target_add_ram32(target, RAM_START, ram_size);
+	target_flash_s *flash = calloc(1, sizeof(*flash));
+	if (!flash) { /* calloc failed: heap exhaustion */
+		DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		return false;
+	}
+	flash->start = FLASH_START;
+	flash->length = flash_size;
+	flash->blocksize = FLASH_PAGE_SIZE;
+	flash->writesize = FLASH_PAGE_SIZE;
+	flash->erase = puya_flash_erase;
+	flash->write = puya_flash_write;
+	flash->prepare = puya_flash_prepare;
+	flash->done = puya_flash_done;
+	flash->erased = 0xffU;
+	target_add_flash(target, flash);
+
+	return false;
+}
+
+static bool puya_flash_prepare(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY1);
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY2);
+
+	int hsi_fs = (target_mem32_read32(flash->t, RCC_ICSCR) >> 13) & 7;
```

Comment:
The masking operation and shifting operation should be unsigned (suffix the constants with `U`), making the result a `uint8_t` - signed bitwise operations are a very prickly part of C that are best avoided.